## Eliminación de Gauss

Es el método más conocido para resolver sistemas de ecuaciones lineales y es considerada la base de los demás métodos de eliminación. Consiste de dos etapas: eliminación y sustitución hacia atrás.

In [74]:
import numpy as np  

A = np.array([
    [0, 7, -1, 3, 1],
    [2, 3, 4, 1, 7],
    [6, 2, 0, 2, -1],
    [2, 1, 2, 0, 2],
    [3, 4, 1, -2, 1]
], float)

b = np.array([5, 7, 2, 3, 4], float)
n = len(b)
x = np.zeros(n, float)
A.shape

(5, 5)

La solucion analitica es x = {0.4444, 0.555, 0.666, 0.222, 0.222}

**Gauss Elimination**

In [75]:
# Elimination
for k in range(n-1):
    if A[k, k] == 0:
        for j in range(n):
            A[k, j], A[k+1, j] = A[k+1, j], A[k, j] 
        b[k], b[k+1] = b[k+1], b[k]
    for i in range(k+1, n):
        if A[i, k] == 0: continue
        factor = A[k, k] / A[i, k] 
        b[i] = b[k] - factor * b[i]       
        for j in range(k, n):
            A[i, j] = A[k, j] - factor * A[i, j]        
print(A)
print(b)

[[ 2.00000000e+00  3.00000000e+00  4.00000000e+00  1.00000000e+00
   7.00000000e+00]
 [ 0.00000000e+00  7.00000000e+00 -1.00000000e+00  3.00000000e+00
   1.00000000e+00]
 [ 0.00000000e+00  0.00000000e+00 -1.30000000e+01  2.00000000e+00
  -2.10000000e+01]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  2.81250000e+00
   5.81250000e+00]
 [ 0.00000000e+00  8.88178420e-16  0.00000000e+00 -4.44089210e-16
   4.95734797e+00]]
[  7.           5.         -14.           0.625        0.15371622]


**Back-Substitution**

In [76]:
# Back-substitution
x[n-1] = b[n-1] / A[n-1, n-1]
for i in range(n-2, -1, -1):
    terms = 0
    for j in range(i+1, n):
        terms += A[i, j] * x[j]
    x[i] = (b[i] - terms) / A[i, i]

In [77]:
print(f"Solution of the System: \n{x}")

Solution of the System: 
[0.02170543 0.79224806 1.05116279 0.15813953 0.03100775]


## Método de Jacobi

Solucion analitica: ${(0.365, -0.233, 0.285, -0.2036)}$ 

In [88]:
import numpy as np
A = np.array([
    [4, 1, 2, -1],
    [3, 6, -1, 2],
    [2, -1, 5, -3],
    [4, 1, -3, -8]
])
b = np.array([2, -1, 3, 2])
(n,) = np.shape(b)
x = np.full(n, 1.0, float) # Initial value of x
xnew = np.empty(n, float) 
iterlimit = 100
tolerance = 1.0e-6
# Iterations
for iteration in range(1, iterlimit+1):
    for i in range(n):
        s = 0 
        for j in range(n):
            if j != i:
                s += A[i, j] * x[j] 
        xnew[i] = (-1/A[i, i]) * (s - b[i])
    if np.all(abs(xnew - x) < tolerance):
        break
    else:
        x = xnew.copy()
print(f"Solution of the System: \n{xnew}")

Solution of the System: 
[ 0.36500668 -0.23378501  0.28506797 -0.20362037]


## **Método de Gauss-Seidel**

In [8]:
import numpy as np
A = np.array([
    [4, 1, 2, -1],
    [3, 6, -1, 2],
    [2, -1, 5, -3],
    [4, 1, -3, -8]
])
b = np.array([2, -1, 3, 2])
(n,) = np.shape(b)
x = np.full(n, 1.0, float) # Initial value of x
xdiff = np.empty(n, float) 
iterlimit = 100
tolerance = 1.0e-6
# Iterations
for iteration in range(1, iterlimit+1):
    for i in range(n):
        s = 0 
        for j in range(n):
            if j != i:
                s += A[i, j] * x[j] 
        xnew = (-1/A[i, i]) * (s - b[i])
        xdiff = abs(xnew - x[i])
        x[i] = xnew
    if np.all(xdiff < tolerance):
        break    
print(f"Solution of the System: \n{x}")

Solution of the System: 
[ 0.36500739 -0.23378566  0.28506799 -0.20362001]


### **Dominio Diagonal**

Antes de aplicar el método de Jacobi o el método de Gauss-Seidel, es importante que se cumpla una condición de convergencia. Es conocida como el *Dominio Diagonal* donde el valor absoluto de los elementos en la diagonal principal debe ser mayor que los demás elementos de su fila.

Ejemplo de no tener Dominio Diagonal:

Solución analítica: $(0.3650, -0.2337, 0.2850, -0.2036)$

In [10]:
A = np.array([
    [2, -1, 5, -3],
    [4, 1, 2, -1],
    [4, 1, -3, -8],
    [3, 6, -1, 2]
]) 
b = np.array([3, 2, 2, -1])
# Aplicamos Gauss-Seidel
(n,) = np.shape(b)
x = np.full(n, 1.0, float) # Initial value of x
xdiff = np.empty(n, float) 
iterlimit = 100
tolerance = 1.0e-6
# Iterations
for iteration in range(1, iterlimit+1):
    for i in range(n):
        s = 0 
        for j in range(n):
            if j != i:
                s += A[i, j] * x[j] 
        xnew = (-1/A[i, i]) * (s - b[i])
        xdiff = abs(xnew - x[i])
        x[i] = xnew
    if np.all(xdiff < tolerance):
        break    
print(f"La solucion no converge a la solucion Analitica: \n{x}")

La solucion no converge a la solucion Analitica: 
[ 8.43572172e+109 -2.71345274e+110 -1.09039090e+110  6.32980450e+110]


## **Linear System Solution in Numpy and SciPy**

Solución analítica: $(0.3650, -0.2337, 0.2850, -0.2036)$

In [15]:
import numpy as np
A = np.array([
    [2, -1, 5, -3],
    [4, 1, 2, -1],
    [4, 1, -3, -8],
    [3, 6, -1, 2]
], float) 
b = np.array([3, 2, 2, -1], float)

from scipy.linalg import solve, inv
x = solve(A, b)
print(x)

[ 0.36500754 -0.23378582  0.28506787 -0.20361991]


In [16]:
x = np.dot(inv(A), b)  
print(x)

[ 0.36500754 -0.23378582  0.28506787 -0.20361991]


In [20]:
x = np.linalg.solve(A, b)
print(x)

[ 0.36500754 -0.23378582  0.28506787 -0.20361991]
